File: Kroeg.Server/Services/Template/TemplateService.cs
﻿using Kroeg.Server.Models;
using Kroeg.Server.Services.EntityStore;
using Kroeg.Server.Tools;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;

namespace Kroeg.Server.Services.Template
{
    public class TemplateService
    {
        public Dictionary<string, List<TemplateItem>> Templates { get; } = new Dictionary<string, List<TemplateItem>>();

        private string _base = "templates/";
        private EntityData _entityData;

        public string PageTemplate { get; private set; }

        public class Registers {
            internal List<string> Load { get; set; }
            internal int Accumulator { get; set; }

            public Dictionary<string, APEntity> UsedEntities { get; set; } = new Dictionary<string, APEntity>();
        }

        public TemplateService(EntityData entityData)
        {
            _entityData = entityData;
            _parse(_base);
        }

        private void _parse(string dir)
        {
            foreach (var file in Directory.EnumerateFiles(dir))
                _parseFile(file);

            foreach (var subdir in Directory.EnumerateDirectories(dir))
                _parse(subdir);
        }

        private void _parseFile(string path)
        {
            var data = File.ReadAllText(path);
            if (path == "templates/page.html") PageTemplate = data;
            Templates[path.Substring(_base.Length, path.Length - _base.Length - 5).Replace('\\', '/')] = TemplateParser.Parse(data);
        }

        private async Task<bool> _parseCondition(APEntity entity, IEntityStore entityStore, string text, Registers regs)
        {
            return await _parseCommand(entity, entityStore, text, regs) != null;
        }

        private async Task<string> _parseCommand(APEntity entity, IEntityStore entityStore, string command, Registers regs)
        {
            JToken _inbetween = null;
            bool isHtml = false;
            var splitCommand = command.Split(' ');
            for (var i = 0; i < splitCommand.Length; i++)
            {
                var asf = splitCommand[i];
                var isAnd = false;
                var previousResult = _inbetween;
                if (asf.StartsWith("&"))
                {
                    isAnd = true;
                    asf = asf.Substring(1);
                }

                if (asf.StartsWith("$"))
                {
                    var name = asf.Substring(1);
                    if (_inbetween != null) continue;
                    var obj = new List<JToken>();
                    foreach (var item in entity.Data[name])
                        if (item.SubObject == null)
                            obj.Add(JToken.FromObject(item.Primitive));
                        else
                            obj.Add(item.SubObject.Serialize(false));
                    if (obj.Count == 0) _inbetween = null;
                    else _inbetween = new JArray(obj.ToArray());
                }
                else if (asf.StartsWith("%"))
                {
                    if (_inbetween == null) continue;
                    string id;
                    if (_inbetween.Type == JTokenType.Array)
                        id = _inbetween[0].ToObject<string>();
                    else
                        id = _inbetween.ToObject<string>();
                    var toCheck = await entityStore.GetEntity(id, true);
                    regs.UsedEntities[id] = toCheck;
                    var obj = new List<JToken>();
                    foreach (var item in toCheck.Data[asf.Substring(1)])
                        if (item.SubObject == null)
                            obj.Add(JToken.FromObject(item.Primitive));
                        else
                            obj.Add(item.SubObject.Serialize(false));
                    if (obj.Count == 0) _inbetween = null;
                    else _inbetween = new JArray(obj.ToArray());
                }
                else if (asf.StartsWith("'"))
                    _inbetween = _inbetween ?? asf.Substring(1);
                else if (asf == "ishtml")
                    isHtml = true;
                else if (asf.StartsWith("render:"))
                {
                    var template = asf.Substring("render:".Length);
                    if (_inbetween == null) return await ParseTemplate(template, entityStore, entity, regs);
                    string id;
                    if (_inbetween.Type == JTokenType.Array)
                        _inbetween = _inbetween[0];

                    if (_inbetween.Type == JTokenType.Object)
                        id = _inbetween["id"].ToObject<string>();
                    else
                        id = _inbetween.ToObject<string>();
                    var newEntity = await entityStore.GetEntity(id, true);
                    regs.UsedEntities[id] = newEntity;
                    return await ParseTemplate(template, entityStore, newEntity, regs);
                }
                else if (asf == "load")
                {
                    regs.Load = new List<string>();
                    if (_inbetween != null)
                    foreach (var item in (JArray) _inbetween)
                    {
                        regs.Load.Add(item.Value<string>());
                    }
                    regs.Accumulator = -1;
                    _inbetween = null;
                }
                else if (asf == "item")
                {
                    if (_inbetween == null && regs.Accumulator >= 0 && regs.Accumulator < regs.Load.Count) _inbetween = regs.Load[regs.Accumulator];
                }
                else if (asf == "next")
                {
                    regs.Accumulator++;
                    _inbetween = (regs.Accumulator < regs.Load.Count) ? regs.Load[regs.Accumulator] : null;
                }
                else if (asf == "server")
                {
                    _inbetween = "server";
                }
                else if (asf == "is")
                {
                    var test = splitCommand[++i];
                    if (test == "Activity") _inbetween = _entityData.IsActivity(entity.Data) ? "Activity" : _inbetween;
                    else if (test == "Collection") _inbetween = entity.Data["type"].Any((a) => (string)a.Primitive == "Collection" || (string)a.Primitive == "OrderedCollection") ? "Collection" : _inbetween;
                    else if (test == "CollectionPage") _inbetween = entity.Data["type"].Any((a) => (string)a.Primitive == "CollectionPage" || (string)a.Primitive == "OrderedCollectionPage") ? "CollectionPage" : _inbetween;
                    else _inbetween = entity.Data["type"].Any((a) => (string)a.Primitive == test) ? test : _inbetween;
                }

                if (isAnd)
                    _inbetween = previousResult == null ? null : _inbetween;
            }

            if (_inbetween == null) return null;

            string text;
            if (_inbetween.Type == JTokenType.Array) text = _inbetween[0].ToObject<string>();
            else text = _inbetween.ToObject<string>();

            if (isHtml) return text;
            return WebUtility.HtmlEncode(text);
        }

        public async Task<string> ParseTemplate(string template, IEntityStore entityStore, APEntity entity, Registers regs = null)
        {
            if (regs == null) regs = new Registers();
            if (!Templates.ContainsKey(template)) throw new InvalidOperationException($"Template {template} does not exist!");
            var templ = Templates[template];
            var builder = new StringBuilder();
            var end = "";

            for (int i = 0; i < templ.Count; i++)
            {
                var item = templ[i];
                switch (item.Type)
                {
                    case "wrap":
                        builder.Append($"<{item.Data} data-id=\"{entity.Id}\" data-template=\"{template}\">");
                        end = $"</{item.Data.Split(' ')[0]}>";
                        break;
                    case "text":
                        builder.Append(item.Data);
                        break;
                    case "if":
                    case "while":
                        if (!await _parseCondition(entity, entityStore, item.Data.Split(new[] { ' ' }, 2)[1], regs))
                            i = item.Offset - 1;
                        break;
                    case "jump":
                        i = item.Offset - 1;
                        break;
                    case "end":
                        var begin = templ[item.Offset];
                        if (begin.Type == "while")
                            i = item.Offset - 1;
                        break;
                    case "command":
                        if (item.Data.Contains("%")) {
                            builder.Append("<span>");
                        }
                        builder.Append(await _parseCommand(entity, entityStore, item.Data, regs) ?? "");
                        if (item.Data.Contains("%")) {
                            builder.Append("</span>");
                        }
                        break;
                }
            }

            builder.Append(end);

            return builder.ToString();
        }
    }
}

File: Kroeg.Client/js/TemplateService.ts
import { EntityStore } from "./EntityStore";
import * as AS from "./AS";

export class TemplateItem {
    public type: "command" | "if" | "while" | "else" | "jump" | "end" | "text" | "wrap";
    public data: string;
    public offset: number;
}

export type Template = TemplateItem[];

export class TemplateService {
    private _templatePromise: Promise<{[item: string]: Template}>

    constructor() {
        this._templatePromise = this._getTemplates();
    }

    private async _getTemplates(): Promise<{[item: string]: Template}> {
        const result = await fetch("http://localhost:5000/settings/templates");
        return await result.json();
    }

    public getTemplates(): Promise<{[item: string]: Template}> {
        return this._templatePromise;
    }
}

export class RenderResult {
    public result: string[] = [];
    public subRender: {id: string, template: string}[] = [];
}

class Registers {
    public load: string[];
    public accumulator: number;
}

export class TemplateRenderer {
    private templates: {[item: string]: Template};

    public constructor(private templateService: TemplateService, private entityStore: EntityStore) {
    }

    public async prepare() {
        this.templates = await this.templateService.getTemplates();
    }

    public getWrap(template: string): string {
        if (template.startsWith("{")) {
            return JSON.parse(template).wrap;
        }
        if (!(template in this.templates)) return null;
        if (this.templates[template][0].type != "wrap") return null;
        return this.templates[template][0].data;
    }

    private _parseCondition(object: AS.ASObject, text: string, reg: Registers): boolean {
        if (text == "next") {
            reg.accumulator++;
            return reg.accumulator < reg.load.length;
        } else if (text == "client") return true;

        const split = text.split(' ');
        if (split.length == 2) {
            if (text == "is Activity") return "actor" in object;
            else if (text == "is Collection") return AS.containsAny(object, "type", ["Collection", "OrderedCollection"]);
            else if (text == "is CollectionPage") return AS.containsAny(object, "type", ["CollectionPage", "OrderedCollectionPage"]);
            else if (split[0] == "is") return AS.contains(object, "type", split[1]);
            else if (split[0] == "has") return AS.get(object, split[1]).length > 0;
            return false;
        }

        const value = split[0];
        const arr = AS.get(object, split[2]);

        switch (split[1]) {
        case "in":
            return arr.indexOf(value) != -1;
        }

        return false;
    }

    private _parseCommand(object: AS.ASObject, command: string, renderResult: RenderResult, reg: Registers): string|{template: string}
    {
        let result: any = null;
        let isHtml = false;
        let depend: string = null;
        if (command.indexOf('%') != -1) {
            let nodepend = command.split(' %', 1)[0];
            depend = "$" + command.substring(command.indexOf('%') + 1);
            command = nodepend;
        }
        for (let asf of command.split(' ')) {
            if (asf.startsWith("$"))
            {
                if (result !== null) continue;
                const name = asf.substring(1);
                let results = [];
                for (let item of AS.get(object, name)) {
                    if ((typeof item) == "object" && !Array.isArray(item)) results.push(JSON.stringify(item));
                    else results.push(item);
                }

                if (results.length == 0) result = null;
                else result = results;
            } else if (asf.startsWith("'")) {
                if (result === null) result = asf.substring(1);
            } else if (asf == "ishtml") {
                isHtml = true;
            } else if (asf.startsWith("render:")) {
                const template = asf.substring(7);
                if (result == null) {
                    return {template};
                }
                let id: string = null;
                if (Array.isArray(result))
                    id = result[0] as string;
                else id = result as string;

                renderResult.subRender.push({id, template});
                return null;
            } else if (asf == "load") {
                reg.load = result as string[];
                reg.accumulator = -1;
                result = null;
            } else if (asf == "item") {
                if (reg.accumulator < reg.load.length && reg.accumulator >= 0 && result == null) result = reg.load[reg.accumulator];
            } else if (asf.startsWith("client.")) {
                if (asf == "client.stats") result = `Preloaded ${Object.keys((window as any).preload).length} items`;
            }
        }

        if (depend != null) {
            if (Array.isArray(result)) result = result[0];
            renderResult.subRender.push({template: JSON.stringify({command: depend, wrap: "span"}), id: result});
            return null;
        }

        if (result == null) return "";

        let text: string;
        if (Array.isArray(result)) text = result[0].toString();
        else text = result == null ? "" : result.toString();

        if (!isHtml) text = text.replace(/</g, "&lt;").replace(/>/g, "&gt;");;

        return text;
    }
    
    public render(template: string, object: AS.ASObject, renderResult?: RenderResult): RenderResult {
        if (renderResult == null) renderResult = new RenderResult();
        if (template.startsWith('{')) {
            // pseudo-template!
            let data = JSON.parse(template);
            let parsed = this._parseCommand(object, data.command, renderResult, new Registers());
            renderResult.result.push(parsed as string);
            return renderResult;
        }

        let temp = this.templates[template];
        let result = "";
        let reg = new Registers();
        for (let i = 0; i < temp.length; i++) {
            const item = temp[i];
            switch (item.type) {
                case "text":
                    result += item.data;
                    break;
                case "if":
                case "while":
                    if (!this._parseCondition(object, item.data.substring(item.data.indexOf(' ') + 1), reg)) i = item.offset - 1;
                    break;
                case "jump":
                    i = item.offset - 1;
                    break;
                case "end":
                    if (temp[item.offset].type == "while")
                        i = item.offset - 1;
                    break;
                case "command":
                    let parsed = this._parseCommand(object, item.data, renderResult, reg);
                    if (parsed != null && typeof parsed == "object") {
                        let template = (parsed as {template: string}).template;
                        let offset = renderResult.result.length;
                        result += `<${this.templates[template][0].data}>`;
                        this.render(template, object, renderResult);
                        renderResult.result[offset] = result + renderResult.result[offset];
                        result = renderResult.result[renderResult.result.length - 1] + `</${this.templates[template][0].data.split(' ')[0]}>`;
                        renderResult.result.splice(renderResult.result.length - 1);
                    } else if (parsed == null) {
                        renderResult.result.push(result);
                        result = "";
                    } else {
                        result += parsed;
                    }

                    break;
            }
        }
        renderResult.result.push(result);
        return renderResult;
    }
}